File: rust/src/euler081.rs
// Problem 81: Path Sum: Two Ways
//
// In the 5 by 5 matrix below, the minimal path sum from the top left to the bottom right, by only moving right and down, is indicated in bold red and is equal to 2427.
//
// 131	673	234	103	18
// 201	96	342	965	150
// 630	803	746	422	111
// 537	699	497	121	956
// 805	732	524	37	331
//
// Find the minimal path sum from the top left to the bottom right by only moving right and down in matrix.txt, a 31K text file containing an 80 by 80 matrix.
//
// Answer: 427337
//
// Execution time: ~1ms on modern hardware
//
// Optimizations: Dynamic programming on grid
// Interesting notes: Reads matrix from file, uses DP to compute min path

pub fn path_sum_two_ways() -> u64 {
    let content = std::fs::read_to_string("../cpp/src/p081_matrix.txt").unwrap();
    let matrix: Vec<Vec<u64>> = content.lines().map(|line| {
        line.split(',').map(|s| s.trim().parse().unwrap()).collect()
    }).collect();
    let rows = matrix.len();
    let cols = matrix[0].len();
    let mut dp = matrix.clone();
    for i in 0..rows {
        for j in 0..cols {
            if i == 0 && j == 0 {
                continue;
            }
            let mut min = u64::MAX;
            if i > 0 {
                min = min.min(dp[i-1][j]);
            }
            if j > 0 {
                min = min.min(dp[i][j-1]);
            }
            dp[i][j] += min;
        }
    }
    dp[rows-1][cols-1]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn euler_081() {
        assert_eq!(path_sum_two_ways(), 427337);
    }
}

File: python/src/euler081.py
# Problem 81: Path Sum: Two Ways
# In the 5 by 5 matrix below, the minimal path sum from the top left to the bottom right, by only moving to the right and down, is indicated in bold red and is equal to 2427.
# Find the minimal path sum from the top left to the bottom right by only moving right and down in matrix.txt, a 31K text file containing an 80 by 80 matrix.
# Answer: 427337
# Execution time: ~10ms on modern hardware
# Optimizations: Dynamic programming on the matrix
# Interesting notes: AI-generated solution using DP for grid path minimization

def solve():
    with open('../cpp/src/p081_matrix.txt') as f:
        matrix = [list(map(int, line.split(','))) for line in f]
    n = len(matrix)
    dp = [[0] * n for _ in range(n)]
    dp[0][0] = matrix[0][0]
    for i in range(1, n):
        dp[i][0] = dp[i-1][0] + matrix[i][0]
        dp[0][i] = dp[0][i-1] + matrix[0][i]
    for i in range(1, n):
        for j in range(1, n):
            dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + matrix[i][j]
    return dp[n-1][n-1]